builder/builder.go

```diff
@@ -201,83 +201,58 @@ func (b *Builder) SignManifestMoM() error {
-	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
-	}
-	bundles := b.Bundledir
-	if _, err := os.Stat(bundles); os.IsNotExist(err) {
-		clrbundles := "clr-bundles/clr-bundles-" + ver + "/bundles/"
-		if err = os.Mkdir(bundles, 0777); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+	bundleDir := getUpstreamBundlesDir(ver)
+
+	// Clear out other bundle dirs if needed
+	if prune {
+		files, err := ioutil.ReadDir(upstreamBundlesBaseDir)
+		if err != nil {
+			return errors.Wrap(err, "Failed to read bundles for pruning.")
 		}
-		// Copy all bundles over into mix-bundles if -all passed
-		if allbundles == true {
-			files, err := ioutil.ReadDir("clr-bundles/clr-bundles-" + ver + "/bundles/")
-			if err != nil {
-				helpers.PrintError(err)
-				os.Exit(1)
-			}
-			for _, file := range files {
-				if err = helpers.CopyFile(bundles+"/"+file.Name(), clrbundles+file.Name()); err != nil {
-					helpers.PrintError(err)
+		for _, file := range files {
+			// Skip the current version cache, but delete all others
+			if strings.Compare(file.Name(), fmt.Sprintf(upstreamBundlesVerDir, ver)) != 0 {
+				if err = os.RemoveAll(filepath.Join(upstreamBundlesBaseDir, file.Name())); err != nil {
+					return errors.Wrapf(err, "Failed to remove %s while pruning bundles", file.Name())
 				}
 			}
-		} else {
-			// Install only a minimal set of bundles
-			fmt.Println("Adding os-core, os-core-update, kernel-native, bootloader to mix-bundles...")
-			_ = helpers.CopyFile(bundles+"/os-core", clrbundles+"os-core")
-			_ = helpers.CopyFile(bundles+"/os-core-update", clrbundles+"os-core-update")
-			_ = helpers.CopyFile(bundles+"/kernel-native", clrbundles+"kernel-native")
-			_ = helpers.CopyFile(bundles+"/bootloader", clrbundles+"bootloader")
 		}
+	}
 
-		// Save current dir so we can get back to it
-		curr, err := os.Getwd()
-		if err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
-		}
-		if err = os.Chdir(b.Bundledir); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
-		}
-		helpers.Git("init")
-		helpers.Git("add", ".")
-		commitMsg := fmt.Sprintf("Initial Mix Version %s from Clear Version %s", b.Mixver, b.Clearver)
-		helpers.Git("commit", "-m", commitMsg)
-		if err = os.Chdir(curr); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
-		}
+	// Check if bundles already exist
+	if _, err := os.Stat(bundleDir); err == nil {
+		return nil
+	}
+
+	tmptarfile := filepath.Join(upstreamBundlesBaseDir, ver+".tar.gz")
+	URL := "https://github.com/clearlinux/clr-bundles/archive/" + ver + ".tar.gz"
+	if err := helpers.Download(tmptarfile, URL); err != nil {
+		return errors.Wrapf(err, "Failed to download bundles for upstream version %s", ver)
 	}
 
-	fmt.Println("Downloaded " + repo)
+	if err := helpers.UnpackFile(tmptarfile, upstreamBundlesBaseDir); err != nil {
+		return errors.Wrapf(err, "Error unpacking bundles for upstream version %s\n%s left for debuging", ver, tmptarfile)
```

Comment:
If Unpack fails, you probably want to remove the directory, otherwise it might leave a partial unpack and your cache check some lines above think everything is fine in a second execution. I'm mentioning this because this kind of issue already happened in Mixer in the past.

Suggestion: you are caching the uncompressed contents. Cache only the tarballs instead, tar+gz is cheap and it might give you less moving parts to care about in your cache (one file vs a lot of stuff).

Reply:
>If Unpack fails, you probably want to remove the directory, otherwise it might leave a partial unpack and your cache check some lines above think everything is fine in a second execution. I'm mentioning this because this kind of issue already happened in Mixer in the past.

Great catch. Fixed in upcoming push.

>Suggestion: you are caching the uncompressed contents. Cache only the tarballs instead, tar+gz is cheap and it might give you less moving parts to care about in your cache (one file vs a lot of stuff).

I considered that, but I decided that
1. I'd rather not have to untar+gz for each command, even if it's cheap. Just feels like unnecessary work to do each time.
1. While they're uncompressed, they're small text files, so compression isn't buying us a lot of savings from that perspective.
1. It's a _little bit_ more future-proof to rely on the bundle output itself, rather than the way mainline CLR currently serves up its bundle definitions as a .tgz. This isn't a great argument, 'cause it's still pretty darn hard-coded to the way CLR does it right now. As I mentioned elsewhere, we have a problem for derivatives of CLR: the way we fetch upstream definitions isn't generic/hot-swappable for something else. We could parameterize the URL we fetch, but that still locks derivatives into serving their bundle definitions as .tgz's with the same internal file structure as mainline CLR.